lib/Refactoring/AddExplicitCodableImplementation.cpp

```diff
@@ -12,183 +12,258 @@
 
 using namespace swift::refactoring;
 
 namespace {
 class AddCodableContext {
 
   /// Declaration context
-  DeclContext *DC;
+  IterableDeclContext *IDC;
 
-  /// Start location of declaration context brace
-  SourceLoc StartLoc;
+  AddCodableContext(IterableDeclContext *IDC) : IDC(IDC){};
+  AddCodableContext(nullptr_t) : IDC(nullptr){};
 
-  /// Array of all conformed protocols
-  SmallVector<swift::ProtocolDecl *, 2> Protocols;
-
-  /// Range of internal members in declaration
-  DeclRange Range;
-
-  bool conformsToCodableProtocol() {
-    for (ProtocolDecl *Protocol : Protocols) {
-      if (Protocol->getKnownProtocolKind() == KnownProtocolKind::Encodable ||
-          Protocol->getKnownProtocolKind() == KnownProtocolKind::Decodable) {
-        return true;
-      }
+  const NominalTypeDecl *getNominal() const {
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal;
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getExtendedNominal();
     }
-    return false;
-  }
 
-public:
-  AddCodableContext(NominalTypeDecl *Decl)
-      : DC(Decl), StartLoc(Decl->getBraces().Start),
-        Protocols(getAllProtocols(Decl)), Range(Decl->getMembers()){};
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  AddCodableContext(ExtensionDecl *Decl)
-      : DC(Decl), StartLoc(Decl->getBraces().Start),
-        Protocols(getAllProtocols(Decl->getExtendedNominal())),
-        Range(Decl->getMembers()){};
+  AccessLevel getContainerAccess() const {
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal->getFormalAccess();
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getDefaultAccessLevel();
+    }
 
-  AddCodableContext() : DC(nullptr), Protocols(), Range(nullptr, nullptr){};
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  static AddCodableContext
-  getDeclarationContextFromInfo(ResolvedCursorInfoPtr Info);
+  /// Get the left brace location of the type-or-extension decl.
+  SourceLoc getLeftBraceLoc() const {
+    if (auto *nominal = dyn_cast<NominalTypeDecl>(IDC->getDecl())) {
+      return nominal->getBraces().Start;
+    }
+    if (auto *extension = dyn_cast<ExtensionDecl>(IDC->getDecl())) {
+      return extension->getBraces().Start;
+    }
 
-  void printInsertionText(ResolvedCursorInfoPtr CursorInfo, SourceManager &SM,
-                          llvm::raw_ostream &OS);
+    llvm_unreachable("unknown IterableDeclContext");
+  }
 
-  bool isValid() { return StartLoc.isValid() && conformsToCodableProtocol(); }
+  /// Get the token location where the text should be inserted after.
+  SourceLoc getInsertStartLoc() const {
+    // Prefer the end of elements.
+    for (auto *member : llvm::reverse(IDC->getParsedMembers())) {
+      if (isa<AccessorDecl>(member) || isa<VarDecl>(member))
+        continue;
```

Comment:
Is this needed because what we really care about are the `PatternBindingDecl` and the `AccessorDecl` and `VarDecl` are only in the `IterableDeclContext` because of the way the AST is modeled?

Reply:
Right, added a comment to clarify it.